### ISSUE TITLE
[3.3.5] Creature attack pattern while rooted is incorrect

### ISSUE THREAD

**Issue by Treeston** [TrinityCore/TrinityCore#15496]
Branch 3.3.5, revision 3f3ff8b36b05, TDB 335.59.

While a creature is rooted (entangling roots, frost nova etc) and the highest threat target is not in melee range, the creature just stands facing the highest threat target and does not attack.

Expected behavior is to have the creature attack the highest threat target within its melee range instead.

**Comment by tkrokli** [comment]
While your issue might point to `StorageType threatlist = me->getThreatManager().getThreatList();` or subsequent procedure or referencing code, I recommend that you update your core (your core rev. is more than a month old) to make sure your files are current enough for editing in any PR you may want to create.

As for the expected creature behaviour, have you taken into consideration if the creature is a caster class or not, as well as if any form of silencing occurs or not on a caster class creature?

**Comment by MrSmite** [comment]
On retail in WotLK, creatures didn't switch targets unless another target generated more threat than their current target. Just because a high-threat target is unreachable doesn't mean the creature gives up.

In a raid situation, a tank would build up aggro and then someone would CC that target while the tank went elsewhere. When the CC broke, the target would automatically pick up the tank again. In your scenario it would start attacking anyone thereby ignoring the threat of the tank.

**Comment by Treeston** [comment]
Crowd Control (Polymorph et al) is different from immobilization (Frost Nova etc).

Both will maintain the current threat table, but while the creature is unable to attack during crowd control effects, it can still attack while immobilized, and should always attack the highest threat target it is capable of reaching - thus, if the creature is melee, it should attack the highest threat adjacent target if it cannot reach its current victim while immobilized.

**Comment by Treeston** [comment]
@tkrokli this is still happening as of c5cc7cca3e29e176b5ec0405a05c3db8e19c4e93, by the way.

**Comment by tkrokli** [comment]
Oh. I rarely take time to play much these days, except for testing simple issues alone (single character) on my local computer. My first comment above was mostly meant as food for thought ref your first post.

**Comment by MrSmite** [comment]
@Treeston Do you have any evidence of this? When a creature has a tank as priority 1 on its list and a Mage uses Frost Nova, that creature should not start attacking the nearby Priest / Mage / whatever. It should maintain focus on the tank and run to the tank when the root breaks.

The only time an NPC should switch targets is when another player attacks it and generates more threat or if it is about to return home, it may look for a new target.

Note: This is from my raid experience during WotLK, I don't know how it is in 6.x

**Comment by Treeston** [comment]
Anecdotal evidence, plus [Wowpedia](http://wow.gamepedia.com/Root) lists this behavior too, and has done so since before WotLK.
>Note that if a mob with no ranged attacks is rooted from range, it will be unable to attack its current aggro target. However, if players are standing in melee range, it will attack the player that has the highest threat level. Thus, it is not wise to stand near a rooted mob, or root one within a crowd of players without prior warning.

**Comment by StormBytePP** [comment]
I am not sure of this one. Suppose you aggro a creature and rot it, then, someone passes by that creature (will generate agro just because of agro radius), should it start attacking the other player just because he went near the mob? I think not.

Furthermore, I never seen in retail a mob you are owning to start attacking other player than you, unless that player generates more aggro than you are generating, even when rooted, but not switching magically targets just because of rooted, in fact, that's a benefit of rooting a mob, making him unable to attack its target until undone.

**Comment by Treeston** [comment]
Proximity should not generate aggro if the creature already is engaged. See unmerged pull #15501, this is a bug with smartAI.

**Comment by Aokromes** [comment]
Sorry, but i must disagree.
If you fight a mob,ie a wolf or one elemental, and nearly is another wolf or elemental, the 2nd wolf or elemental will agro you.
(seen yesterday on retail multiple times)

**Comment by StormBytePP** [comment]
We are talking about the same mob, not nearby mobs, I was talking about nearby players in relation to the same mobs. Anyway, I never seen a mob changing target just because rooted, they follow agro, but I may be wrong, since some time has passed since I do not do quests grouped.

**Comment by Treeston** [comment]
@Aokromes yes, of course. But if someone is fighting an aggressive wolf, and you run past the aggressive wolf, you shouldn't be put on the wolf's aggro table.

**Comment by Treeston** [comment]
@StormBytePP my recollection was that mages were always told to not use nova when they pull aggro because it's gonna make mobs punch other casters.

**Comment by StormBytePP** [comment]
@Treeston Of course, if thouse mobs are not already aggro'd, nova will generate agro enough to get them engaged, if that happens and you lose agro because they generate more than you, then they are attacked. Anyway if mob is a caster the whole of this don't apply as they can still attack on range.

**Comment by Treeston** [comment]
If mage is standing next to priest and mob starts attacking mage, mage panic novas and runs, mob smacks priest even if tank gets aggro back because rooted.

**Comment by Trisjdc** [comment]
Won't comment on whether the behavior is wrong or not, but this should be easy to solve

Just a matter of changing Unit::isInAccessiblePlaceFor. Probably need to add a MotionMaster check too.

**Comment by MrSmite** [comment]
> If mage is standing next to priest and (melee) mob starts attacking mage, mage panic novas and runs, mob smacks priest even if tank gets aggro back because rooted and priest is the only thing in melee range.

Healing generates a lot of threat. I think what you're seeing is other players stealing threat from the tank.

**Comment by Treeston** [comment]
Because you are making me think I'm completely bonkers by now, I've looked for more sources:

> Don't freeze melee adds next to the healer or other ranged DPS because they'll start beating on them.
([Source](http://us.battle.net/wow/en/forum/topic/2593047043#7))

>Frost Nova. While frost nova is a great tool for surviving out in the wild, it's not so great for use in instances indiscriminately. A rooted mob will attack the highest person on his hate list that he can reach. Frost novaing a mob near the healer is a great way to wipe the group. If you get aggro, try to frost nova the mob near the tank, away from all other party members.
([Source](http://www.wowhead.com/forums&topic=13158/threat-and-you#p114504))

> This week and next, we look at the time-honored tradition of mages dying whenever something looks at them funny and discuss a few ways to break that tradition. Way #1: Stand next to the warlock, pull aggro, cast Frost Nova, then Blink away.
([Source](http://www.engadget.com/2011/02/05/arcane-brilliance-the-mage-survival-guide-part-1/))

> You might want to note that Mobs will always try to attack something. So if you frost nova them and you're then out of reach they'll try and attack something else unfriendly, in my case it was the level 17 Warlock I was twinking through Deadmines that had decided to stand in the middle of the group of mobs I was aoe-ing... Not a good sight. This is probably one of the few cases in which mobs will ignore their agro-metres.
([Source](http://www.wowhead.com/spell=122/frost-nova#comments))

> Be careful when using this spell in instances. If you frost nova a group of mobs, and the tank moves away from them, they will hit whoever they can reach if they're melee. This usually means that frost nova kills the rogue.
([Source](http://www.wowhead.com/spell=122/frost-nova#comments:id=284899))

There's probably more, but that's the ones I found in the first few pages of a google search. Please note that all of the above sources are from Wrath, or even earlier.

**Comment by MrSmite** [comment]
@Treeston 

What those comments you linked don't take into account is aggro drop. When the Mage uses Frost Nova he/she becomes first priority. If the Mage moves away and nobody damages the creature for a period, it will drop aggro from the initial target. It is at this point it will attempt to seek another target but it doesn't happen as soon as the initial aggressor is out of range.

**Comment by Keader** [comment]
Still valid after Threat Rewrite @Treeston ?

**Comment by Treeston** [comment]
Yes. I never changed this behavior, as there was no consensus to support my recollection.

**Comment by Shauren** [comment]
I confirm that rooted creatures should switch target to highest threat in melee range
